src/replication/replication_source_db_copy.cpp

```diff
@@ -0,0 +1,697 @@
+    for (const OID class_oid : m_class_oid_list)
+      {
+	heap_extract_worker_task *task = new heap_extract_worker_task (*this, class_oid);
+	inc_extract_running_thread ();
+	cubthread::get_manager ()->push_task (m_heap_extract_workers_pool, task);
+      }
+
+    /* wait for all heap extract threads to end */
+    while (get_extract_running_thread () > 0)
+      {
+	if (is_interrupted (thread_ref))
+	  {
+	    error = ER_INTERRUPTED;
+	    er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error, 0);
+	    return error;
+	  }
+	thread_sleep (10);
+      }
+    execute_and_transit_phase (HEAP_COPY_FINISHED);
+    pack_and_add_end_of_extract_heap ();
+
+    execute_and_transit_phase (SCHEMA_COPY_TRIGGERS);
+    execute_and_transit_phase (SCHEMA_COPY_INDEXES);
+    /* wait for indexes and triggers schema */
+    error = wait_send_triggers_indexes (thread_ref);
+    if (error != NO_ERROR)
+      {
+	return error;
+      }
+
+    pack_and_add_end_of_copy ();
+
+    error = wait_slave_finished ();
+
+    return error;
+  }
+
+  int source_copy_context::setup_copy_protocol (cubcomm::channel &chn)
+  {
+    UINT64 pos = 0, expected_magic;
+    std::size_t max_len = sizeof (UINT64);
+    cubstream::stream_position online_repl_pos;
+
+    if (chn.recv ((char *) &expected_magic, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
+
+    if (expected_magic != replication_node::SETUP_COPY_REPLICATION_MAGIC)
+      {
+	er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::setup_copy_protocol error in setup protocol");
+	assert (false);
+	return ER_FAILED;
+      }
+
+    if (chn.send ((char *) &replication_node::SETUP_COPY_REPLICATION_MAGIC, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
+
+    /* TODO */
+    online_repl_pos = m_online_replication_start_pos;
+    pos = htoni64 (online_repl_pos);
+    if (chn.send ((char *) &pos, max_len) !=  css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
+    er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::setup_copy_protocol online_repl_pos:%lld",
+			      online_repl_pos);
+
+    return NO_ERROR;
+  }
+
+  int source_copy_context::wait_slave_finished ()
+  {
+    bool sender_alive = true;
+
+    er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::wait_slave_finished");
+
+    assert (m_transfer_sender != NULL);
+    assert (m_senders_manager != NULL);
+
+    m_senders_manager->stop_stream_sender (m_transfer_sender);
+
+    while (sender_alive)
+      {
+	sender_alive = m_senders_manager->find_stream_sender (m_transfer_sender);
+	if (sender_alive)
+	  {
+	    thread_sleep (100);
+	  }
+      }
+
+    er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::wait_slave_finished OK");
+
+    return NO_ERROR;
+  }
+
+  void source_copy_context::stop ()
+  {
+    std::unique_lock<std::mutex> ulock_state (m_state_mutex);
+    m_is_stop = true;
+    m_state_cv.notify_all();
```

Comment:
ulock_state may be unlocked before notify

Reply:
fixed